coral/src/app/features/configuration/environments/components/EnvironmentsTabs.test.tsx

```diff
@@ -0,0 +1,223 @@
+import { cleanup, screen } from "@testing-library/react";
+import userEvent from "@testing-library/user-event";
+import EnvironmentsTabs from "src/app/features/configuration/environments/components/EnvironmentsTabs";
+import {
+  ENVIRONMENT_TAB_ID_INTO_PATH,
+  EnvironmentsTabEnum,
+  Routes,
+} from "src/app/router_utils";
+import * as environmentsApi from "src/domain/environment/environment-api";
+import { EnvironmentPaginatedApiResponse } from "src/domain/environment/environment-types";
+import { customRender } from "src/services/test-utils/render-with-wrappers";
+
+class Deferred<T> {
+  public promise: Promise<T>;
+  public resolve!: (value: T | PromiseLike<T>) => void;
+  public reject!: (value: T | PromiseLike<T>) => void;
+  constructor() {
+    this.promise = new Promise((resolve, reject) => {
+      this.resolve = resolve;
+      this.reject = reject;
+    });
+  }
+}
+
+const mockedNavigate = jest.fn();
+jest.mock("react-router-dom", () => ({
+  ...jest.requireActual("react-router-dom"),
+  useNavigate: () => mockedNavigate,
+}));
+
+const mockedKafkaTotalEnvs: EnvironmentPaginatedApiResponse = {
+  totalPages: 1,
+  currentPage: 1,
+  totalEnvs: 1,
+  entries: [
+    {
+      type: "kafka",
+      name: "DEV",
+      id: "1001",
+      params: {},
+    },
+  ],
+};
+
+const mockedSchemaTotalEnvs: EnvironmentPaginatedApiResponse = {
+  totalPages: 1,
+  currentPage: 1,
+  totalEnvs: 2,
+  entries: [
+    {
+      type: "schemaregistry",
+      name: "DEV",
+      id: "1001",
+      params: {},
+    },
+    {
+      type: "schemaregistry",
+      name: "TST",
+      id: "1002",
+      params: {},
+    },
+  ],
+};
+
+const mockedConnectorTotalEnvs: EnvironmentPaginatedApiResponse = {
+  totalPages: 0,
+  currentPage: 0,
+  totalEnvs: 0,
+  entries: [],
+};
+
+describe("EnvironmentsTabs", () => {
+  let user: ReturnType<typeof userEvent.setup>;
```

Comment:
The api mocking in the tests is not correct - you reset the mock afterEach in the first scope, but it's not mocked for every `it` block, so there will be cases where it's `undefined`. Also, the pipeline fails with an error that shows that's there's a memory leak and at least one api call wasn't mocked and the error log for that shows up in the end. Since this is the only test that was touched, it's happening here. It probably won't show up in this tests logs, because it's `Deferred`

Reply:
I think the implementation of these tests is the same as the ones here: https://github.com/Aiven-Open/klaw/blob/299d13eaa35f70ddde375259d90e86341b1eedd9/coral/src/app/features/approvals/components/ApprovalResourceTabs.test.tsx

So probably it should be fixed there too? I'm not sure why the problem with the leak does not appear in that context though.

I am also unsure about these two things:

> you reset the mock afterEach in the first scope, but it's not mocked for every it block, so there will be cases where it's undefined.

How come the test does not fail if this is is the case?

>  Also, the pipeline fails with an error that shows that's there's a memory leak and at least one api call wasn't mocked and the error log for that shows up in the end. Since this is the only test that was touched, it's happening here. It probably won't show up in this tests logs, because it's Deferred

How come I don't see these error logs when runnnig the tests locally? I am still confused by what they mean, and how the calls should be mocked in the test.

Reply:
In `ApprovalResourceTab`, every `render` has all api calls mocked, in this one e.g. line 171, there's:

```
    beforeEach(() => {
      user = userEvent.setup();
      customRender(
        <EnvironmentsTabs currentTab={EnvironmentsTabEnum.KAFKA} />,
        {
          queryClient: true,
          memoryRouter: true,
        }
      );
    });
```

So the component is rendered for every `it`, but in none of those renders the api calls are mocked. So they are calling the `throw Error`. Since the spied on function expects a Promise (either return or reject), there's no handling for a throw Error, so it will not show up right now. If you replace the `throw Error` with returning a Promise.reject() with the same message, it will at least show up in the logs. 

To debug stuff like that, I run each block individual to see where the errors show up and where they don't and try to see what's different. 

In the current state, the test run is finished before the errors bubble up, so they show up only at the end in the pipeline. 

These kind of memory leaks are often really hard to reproduce with jest, because they can get lost if the tests run a bit faster (often on local machines) etc. There is a flag to catch those leaks but in my experiences it's basically useless 😢

Reply:
So, this is not mocking the API reposne because the `beforeEach` in the outer scope is resetting the spy mocks beforehand? https://github.com/Aiven-Open/klaw/pull/1748/commits/f82c4bd6495fc7ff8b81474be9b7dd6e1bf747be

That's what on line 171, right?

But that is the setup in the Approvals tab test? https://github.com/Aiven-Open/klaw/blob/299d13eaa35f70ddde375259d90e86341b1eedd9/coral/src/app/features/approvals/components/ApprovalResourceTabs.test.tsx#L114

Why isn't it a problem there?

Is the Environment tabs test  missing something like this? [a9a0dfa](https://github.com/Aiven-Open/klaw/pull/1748/commits/a9a0dfa400b2f20a9baf96b1f5a1dc18b0e882c7)